homeassistant/components/withings/__init__.py

```diff
@@ -59,52 +82,124 @@ async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
-                },
+    # Normalize the config entry.
+    webhook_id = entry.data.get(CONF_WEBHOOK_ID, webhook.async_generate_id())
+    hass.config_entries.async_update_entry(
+        entry,
+        unique_id=entry.data["token"]["userid"],  # Upgrade existing configs.
+        data={
+            **entry.data,
+            **{
+                const.CLIENT_ID: hass.data[DOMAIN][const.CONFIG][const.CLIENT_ID],
+                const.CLIENT_SECRET: hass.data[DOMAIN][const.CONFIG][
+                    const.CLIENT_SECRET
+                ],
+                const.USE_WEBHOOK: hass.data[DOMAIN][const.CONFIG][const.USE_WEBHOOK],
+                CONF_WEBHOOK_ID: webhook_id,
+                const.CONF_WEBHOOK_URL: entry.data.get(
+                    const.CONF_WEBHOOK_URL, webhook.async_generate_url(hass, webhook_id)
+                ),
             },
-        )
-
-    implementation = await config_entry_oauth2_flow.async_get_config_entry_implementation(
-        hass, entry
+        },
     )
 
-    data_manager = get_data_manager(hass, entry, implementation)
+    data_manager = await async_init_data_manager(hass, entry)
 
-    _LOGGER.debug("Confirming we're authenticated")
+    webhook.async_register(
+        hass,
+        const.DOMAIN,
+        "Withings notify",
+        data_manager.webhook_config.id,
+        async_webhook_handler,
+    )
+
+    _LOGGER.debug("Confirming %s is authenticated to withings", data_manager.profile)
     try:
-        await data_manager.check_authenticated()
+        await data_manager.async_check_authenticated()
     except NotAuthenticatedError:
-        _LOGGER.error(
-            "Withings auth tokens exired for profile %s, remove and re-add the integration",
-            data_manager.profile,
+        # Returning true so async_unload_entry will work later.
+        return True
+
+    async def start_platforms():
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_setup(
+                    entry, BINARY_SENSOR_DOMAIN
+                ),
+                hass.config_entries.async_forward_entry_setup(entry, SENSOR_DOMAIN),
+            ],
+            loop=hass.loop,
         )
-        return False
+        data_manager.start_polling()
 
-    hass.async_create_task(
-        hass.config_entries.async_forward_entry_setup(entry, "sensor")
-    )
+    hass.async_create_task(start_platforms())
 
     return True
 
 
-async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Unload Withings config entry."""
-    return await hass.config_entries.async_forward_entry_unload(entry, "sensor")
+    data_manager = get_data_manager(hass, entry)
+
+    # Background this operation as the data manager can take a few seconds as it unsubscribes
+    # from webhooks.
+    hass.async_create_task(async_remove_data_manager(hass, entry))
```

Comment:
Let's not background it, as we want to make sure it's done when `async_unload_entry` finishes. Also, could we add it to the `gather` below?